actions/concurrent_runner.go

```diff
@@ -0,0 +1,100 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package actions
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const defaultTimeout = time.Second * 10
+
+// ActionResult represents an action result.
+//nolint:unused
+type ActionResult struct {
+	ID             string
+	Name           string
+	Error          error
+	CombinedOutput []byte
+}
+
+// ConcurrentRunner represents concurrent action runner.
+// Action runner is component that can run an Actions.
+//nolint:unused
+type ConcurrentRunner struct {
+	out    chan ActionResult
+	logger logrus.FieldLogger
+
+	actions sync.Map // map[string]Action
+	timeout time.Duration
+}
+
+// NewConcurrentRunner returns new runner.
+// If timeout is 0 it sets to defaultTimeout constant (10sec).
+func NewConcurrentRunner(l logrus.FieldLogger, timeout time.Duration) *ConcurrentRunner {
+	if timeout == 0 {
+		timeout = defaultTimeout
+	}
+
+	return &ConcurrentRunner{
+		logger:  l,
+		timeout: timeout,
+		out:     make(chan ActionResult),
+	}
+}
+
+// Run runs an Action in separate goroutine.
+// When action is ready those output writes to ActionResult channel.
+// You can get all action results with ActionReady() method.
+func (r *ConcurrentRunner) Run(a Action) {
```

Comment:
"Run" is the whole interval from start to finish. We do not wait for the action to finish, so we should call this method `Start`.